Explainer and spec update for padding

xyaoinum opened the pull request:
1. Update the header format, and include the padding algorithm, per discussion in https://github.com/patcg-individual-drafts/topics/issues/183.
2. Reorder the comparator conditions of BrowsingTopic (i.e. "version" first, and then "topic id"). This aligns with the current implementation.

xyaoinum commented:
@domenic PTAL. Thanks! (Thanks for filing the issues giving suggestions to the spec. I will address them later, but this patch is a behavior change and we'd like to add it to the spec first)

@jkarlin  PTAL. Thanks!

domenic commented on spec.bs:
```diff
@@ -325,13 +325,32 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
   <h2 id="topics-for-caller-header">Topics for caller</h2>
   <div algorithm>
-    To <dfn>calculate the topics for caller</dfn>, given a [=topics caller context=] |callerContext|, perform the following steps. They return a list of {{BrowsingTopic}}s.
+    To <dfn>calculate the topics for caller</dfn>, given a [=topics caller context=] |callerContext|, perform the following steps. They return a list of {{BrowsingTopic}}s
```
Nit: the period was good, keep the period :)

xyaoinum replied:
Done.

domenic commented on spec.bs:
```diff
@@ -491,19 +510,83 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+          1. Insert |topicItem| to |topicsInnerList|.
+      1. Let |maxNumberOfEpochs| be 3 (i.e. topics are selected from the last 3 epochs).
+      1. Let |topicMaxLength| be 3 (i.e. [=browsing topics types/topic id=] has maximum 3 digits).
+      1. Let |versionMaxLength| be 13 (i.e. chrome.1:1:11).
```
This "i.e." isn't clarifying to me. Why is 13 related to `chrome.1:1:11`? What does this mean for user agents with long names, like, say, `internetexplorer`?

xyaoinum replied:
Clarified -- switched to rely on the current [=browsing topics types/version=].

domenic commented on spec.bs:
```diff
@@ -491,19 +510,83 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+      <div class="example">
+          One returned topics, and underlying epochs have two different versions:
+
+          <code>t=(1;v=chrome.1:1:2 2), p=P0000000000000000000000000</code>
```
I only see one version here?

xyaoinum replied:
Done. (Removed. As example are less applicable given we'd always expose all versions.)

domenic reviewed (commented):
This seems closely related to my open issue of #183. Although I am glad you are changing the format to be a bit less problematic, I think the core issue remains: the desire to not be explicit about the version association.

Have you considered just always adding the version, to every topic? That makes parsing much easier, and if I understand this correctly, it also would make the padding less necessary. (Although still necessary in some cases.)

xyaoinum commented on spec.bs:
```diff
@@ -491,19 +510,83 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+      1. Let |innerListItemsSeparatorLength| be 1 (i.e. structured fields use a single character (' ') to separate inner list items).
+      1. Let |maxPaddingLength| be <code>|maxNumberOfEpochs| * |topicMaxLength| + |numVersionsInEpochs| * (|versionMaxLength| + |paramsSeparatorLength| + |paramKeyLength|) + (|maxNumberOfEpochs| - 1) * |innerListItemsSeparatorLength|</code>.
+
+      Note: the numbers used in the step above are specific to one of Chrome's implementation. Implementers should use numbers compatible with their configuration.
```
Done. Some are switched to use the configuration params, some are intended to be fixed constant.

xyaoinum commented:
@domenic I switched to always expose the version per your suggestion (although still would like to see Josh's position), and addressed your other comments. PTAL again. Thanks!

domenic commented on spec.bs:
```diff
@@ -491,19 +491,61 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+      1. Let |paramKeyLength| be 1 (i.e. always use a single character parameter key)
+      1. Let |serializedTopicsInnerListLength| be 0.
       1. For each |topic| in |topics|:
+          1. Let |digitsInTopic| be the number of digits in |topic|["{{BrowsingTopic/topic}}"].
```
```suggestion
          1. Let |digitsInTopic| be the number of base-10 digits in |topic|["{{BrowsingTopic/topic}}"].
```

just for extra clarity

xyaoinum replied:
Done.

domenic commented on README.md:
```diff
@@ -71,9 +71,9 @@ The topics will be inferred by the browser. The browser will leverage a classifi
     * The response header will only be honored if the corresponding request included the topics header (or would have included the header if it wasn't empty).
     * The registrable domain used for topic observation is that of the url of the request.
-    * Example request header: `Sec-Browsing-Topics: 123;v=chrome.1:1:2, 2`
+    * Example request header: `Sec-Browsing-Topics: t=(123;v=chrome.1:1:2 2;v=chrome.1:1:2), p=P000000000000000000000000`
```
So, I realized that `:` is not allowed in structured field tokens. Probably you need to change them to be strings instead.

xyaoinum replied:
I think the spec does allow ":" and "."? https://www.rfc-editor.org/rfc/rfc8941.html#section-3.3.4

domenic replied:
Sorry for missing this; you're right!

domenic commented on spec.bs:
```diff
@@ -491,19 +491,61 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+      1. Let |topicMaxLength| be number of digits in the maximum [=browsing topics types/topic id=] (e.g. for <a href="https://github.com/patcg-individual-drafts/topics/blob/main/taxonomy_v1.md">Chrome's initial taxonomy</a>, |topicMaxLength| is 3, as the [=browsing topics types/topic id=] has maximum 3 digits).
+      1. Let |versionMaxLength| be the length of the current [=browsing topics types/version=].
+      1. Let |innerListItemsSeparatorLength| be 1 (i.e. structured fields use a single character (' ') to separate inner list items).
+      1. Let |maxPaddingLength| be <code>|maxNumberOfEpochs| * (|topicMaxLength| + |versionMaxLength| + |paramsSeparatorLength| + |paramKeyLength|) + (|maxNumberOfEpochs| - 1) * |innerListItemsSeparatorLength|</code>.
```
```suggestion
      1. Let |maxPaddingLength| be |maxNumberOfEpochs| * (|topicMaxLength| + |versionMaxLength| + |paramsSeparatorLength| + |paramKeyLength|) + (|maxNumberOfEpochs| - 1) * |innerListItemsSeparatorLength|.
```

xyaoinum replied:
Done.

domenic commented on spec.bs:
```diff
@@ -491,19 +491,61 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+          1. Insert |topicItem| to |topicsInnerList|.
+      1. Let |maxNumberOfEpochs| be 3 (i.e. topics are selected from the last 3 epochs).
+      1. Let |topicMaxLength| be number of digits in the maximum [=browsing topics types/topic id=] (e.g. for <a href="https://github.com/patcg-individual-drafts/topics/blob/main/taxonomy_v1.md">Chrome's initial taxonomy</a>, |topicMaxLength| is 3, as the [=browsing topics types/topic id=] has maximum 3 digits).
+      1. Let |versionMaxLength| be the length of the current [=browsing topics types/version=].
```
What is the "current" one? [=browsing topics types/version=] is only defined given the three inputs _configurationVersion_, _taxonomyVersion_, and _modelVersion_.

Probably the best solution here is to move this definition to https://patcg-individual-drafts.github.io/topics/#terminology-and-types-header to make it clear you're trying to figure out a user-agent-wide maximum. E.g. something like

> The maximum version string length is the maximum possible string length of a [=browsing topics types/version=] that a user agent could possibly generate in a given browsing session.

(or maybe in a given "software release"??)

and then adding an example would be even nicer.

xyaoinum replied:
Done.

domenic commented on spec.bs:
```diff
@@ -491,19 +491,61 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
       1. [=Set a structured field value=] given ([:Sec-Browsing-Topics:], |headerStructuredFields|) in |request|'s [=request/header list=].
 
-    Note: In Chrome's experimentation phase, it will additionally require a valid <a href="https://github.com/GoogleChrome/OriginTrials/blob/gh-pages/explainer.md">Origin Trial</a> token to exist in |initiatorWindow|'s <a data-cite="!HTML#concept-document-window">associated document</a> for the request to be eligible for topics.
+    <div class="note">
```
Possible idea for spec-writing, feel free to disregard: you could try to generate the padding length by first generating an artificial maximum-length header value, then generating a first draft of your real header value, and then subtracting their serialized lengths. It's inefficient, but it's perhaps easier to understand for the reader, and the implementation could do the more efficient thing.

(You could maybe even generate the artificial maximum-length header value ahead of time, in the terminology section or something?)

xyaoinum replied:
I incorporated your first suggestion. (I kept the generation of maximum-length inline as I feel it'd give more context here.)

domenic reviewed (commented):
I appreciate the move to the simpler format. We can continue discussing other ideas in #183.

Current significant blocker is that tokens cannot contain `:`.

domenic commented on spec.bs:
```diff
@@ -45,6 +45,11 @@ spec: html; urlPrefix: https://wicg.github.io/nav-speculation/
+    type: dfn
+        text: Structured Fields Token; url: rfc8941.html#name-tokens
+    type: dfn
+        text: Structured Fields Parameters; url: rfc8941.html#ame-parameters
```
```suggestion
        text: Structured Fields Parameters; url: rfc8941.html#name-parameters
```

domenic commented on spec.bs:
```diff
@@ -491,19 +509,88 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+          1. Insert |topicItem| to |currentTopicsInnerList|.
+      1. If |currentTopicsInnerList| is not empty:
+          1. Let |topicParameters| be an empty [=Structured Fields Parameters=].
+          1. Set |topicParameters|["<code>v</code>"] to a [=Structured Fields Token=] having words |lastVersion|.
```
```suggestion
          1. Set |topicParameters|["<code>v</code>"] to a [=Structured Fields Token=] with value |lastVersion|.
```

domenic commented on spec.bs:
```diff
@@ -491,19 +509,88 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+      1. Let |topicsStructuredFieldsList| be an empty <a href="https://www.rfc-editor.org/rfc/rfc8941.html#name-lists">Structured Fields List</a>.
+      1. Let |currentTopicsInnerList| be an empty <a href="https://www.rfc-editor.org/rfc/rfc8941.html#name-inner-lists">Structured Fields Inner List</a>.
       1. For each |topic| in |topics|:
+          1. If |lastVersion| is null or if |lastVersion| != |topic|["{{BrowsingTopic/version}}"]:
```
```suggestion
          1. If |lastVersion| is null or if |lastVersion| is not |topic|["{{BrowsingTopic/version}}"]:
```

domenic commented on spec.bs:
```diff
@@ -491,19 +509,88 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+          1. If |lastVersion| is null or if |lastVersion| != |topic|["{{BrowsingTopic/version}}"]:
+              1. If |currentTopicsInnerList| is not empty:
+                  1. Let |topicParameters| be an empty [=Structured Fields Parameters=].
+                  1. Set |topicParameters|["<code>v</code>"] to a [=Structured Fields Token=] having words |lastVersion|.
```
```suggestion
                  1. Set |topicParameters|["<code>v</code>"] to a [=Structured Fields Token=] with value |lastVersion|.
```

domenic commented on spec.bs:
```diff
@@ -491,19 +509,88 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+          1. Increment |paddingLength| by |listItemsSeparatorLength| (i.e. to account for the separator characters that would be added when |topics| are not empty).
+      1. If |paddingLength| < 0, then set |paddingLength| to 0.
+
+      Note: the padding should generally be >= 0. It may be negative in certain circumstances: when historically stored topic versions are greater (and use more digits) than the current [=browsing topics types/maximum version string length=]; or when there is a race between getting topics and getting the number of distinct topic versions. Clamp to 0 to prevent breakage in these rare circumstances ...
```
```suggestion
      Note: the padding should generally be &geq; 0. It may be negative in certain circumstances: when historically stored topic versions are greater (and use more digits) than the current [=browsing topics types/maximum version string length=]; or when there is a race between getting topics and getting the number of distinct topic versions. Clamp to 0 to prevent breakage in these rare circumstances.
```

domenic commented on spec.bs:
```diff
@@ -491,19 +509,88 @@ spec: html; urlPrefix: https://www.rfc-editor.org/rfc/
+      1. Let |paddedToken| be "P".
+      1. Append |paddingLength| <code>"0"</code> characters to the end of |paddedToken|.
+      1. Let |paddedEntryParameters| be an empty [=Structured Fields Parameters=].
+      1. Set |paddedEntryParameters|["<code>p</code>"] to a [=Structured Fields Token=] having words |paddedToken|.
```
```suggestion
      1. Set |paddedEntryParameters|["<code>p</code>"] to a [=Structured Fields Token=] with value |paddedToken|.
```

domenic reviewed (approved):
LGTM with nits and one suggestion for clarity improvement, which you can either do via another review pass or via a followup PR (since I imagine it might be important to get the normative changes settled quickly).

I think the current algorithm is efficient but hard to read. The reuse of currentTopicsInnerList, and the way it relies on the topics that come back being sorted, makes it quite stateful. I'd suggest an easier to read two-pass algorithm like so:

1. Let _versionsToTopics_ be an ordered map.
1. For each _topic_ of _topics_:
    1. Let _version_ be _topic_["`version`"].
    1. Let _topicInteger_ be _topic_["`topic`"].
    1. If _versionsToTopics_[_version_] does not exist, set it to an empty list.
    1. Append _topicInteger_ to _versionsToTopics_[_version_].
1. Let _topicsStructuredFieldsList_ be an empty Structured Fields List.
1. For each _version_ → _topicIntegers_ of _versionsToTopics_:
    1. Let _innerList_ be an empty Structured Fields Inner List.
    1. Append all items from _topicIntegers_ to _innerList_.
    1. Let _topicParameters_ be an empty Structured Fields Parameters.
    1. Set _topicParameters_["`v`"] to a Structured Field Token having value _version_.
    1. Associated _topicParamters_ with _innerList_.
    1. Append _innerList_ to _topicsStructuredFieldsList_.

xyaoinum commented:
Thanks @domenic! Landing this now. I will create followup PR for the clarity improvement and other feedback if any.
